src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -86491,5 +86468,178 @@ public void M(TypedReference r)
+                // (18,9): warning CS8602: Possible dereference of a null reference.
+                //         fa3[0]().ToString(); // warning
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "fa3[0]()").WithLocation(18, 9));
+        }
+
+        [Fact, WorkItem(29617, "https://github.com/dotnet/roslyn/issues/29617")]
+        public void CaptureVariablesWhereLambdaAppears_04()
+        {
+            var source =
+@"using System;
+
+class C
+{
+    static T F<T>(Func<T> f) => throw null!;
+    static void G(object? o)
+    {
+        F((Func<object>)(() => o)).ToString(); // 1
+        if (o == null) return;
+        F((Func<object>)(() => o)).ToString();
+    }
+}
+";
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (8,32): warning CS8603: Possible null reference return.
+                //         F((Func<object>)(() => o)).ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "o").WithLocation(8, 32));
+        }
+
+        [Fact, WorkItem(29617, "https://github.com/dotnet/roslyn/issues/29617")]
+        public void CaptureVariablesWhereLambdaAppears_05()
+        {
+            var source =
+@"using System;
+
+class C
+{
+    static T F<T>(Func<T> f) => throw null!;
+    static void G(Action a) => throw null!;
+    static void M(object? o)
+    {
+        F(() => o).ToString(); // 1
+        if (o == null) return;
+        F(() => o).ToString();
+        G(() => o = null); // does not affect state in caller
+        F(() => o).ToString();
+    }
+}
+";
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (9,9): warning CS8602: Possible dereference of a null reference.
+                //         F(() => o).ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F(() => o)").WithLocation(9, 9));
+        }
+
+        [Fact, WorkItem(29617, "https://github.com/dotnet/roslyn/issues/29617")]
+        public void CaptureVariablesWhereLambdaAppears_06()
+        {
+            var source =
+@"using System;
+
+class C
+{
+    static T F<T>(object? o, params Func<T>[] a) => throw null!;
+    static void M(string? x)
+    {
+        F(x = """", () => x = null, () => x.ToString());
+    }
+}
+";
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(29617, "https://github.com/dotnet/roslyn/issues/29617")]
+        public void CaptureVariablesWhereLambdaAppears_07()
+        {
+            var source =
+@"using System;
+
+class C
+{
+    static T F<T>([System.Runtime.CompilerServices.EnsuresNotNull] object? o, params Func<T>[] a) => throw null!;
+    static void M(string? x)
+    {
+        F(x = """", () => x = null, () => x.ToString());
```

Comment:
> = """" [](start = 11, length = 7)

Is the assignment necessary? Doesn't that mean the `[EnsuresNotNull]` attribute is unnecessary? If so, the test looks equivalent to the previous test.